File: libs/cua-driver/examples/jev-use/experiments/task-timing/task_timing.ts
/** Fork-only timing; fixture observations and runner outcomes are separate facts. */
import { randomUUID } from 'node:crypto';

export type TerminalObservation = {
  verdict: 'verified' | 'refuted';
  at_ms: number;
  source: 'existing_fixture_state_classification';
};
type Event = Record<string, unknown>;

export class TaskTiming {
  readonly taskId = randomUUID();
  readonly clockId = `performance.now:${this.taskId}`;
  private readonly start: number;
  private readonly spans: Event[] = [];
  private dropped = 0;
  private setupEnd: number | null = null;
  private terminal: TerminalObservation | null = null;
  private closed = false;

  constructor(private readonly clock: () => number = () => performance.now(),
              private readonly maxSpans = 512) {
    if (!Number.isInteger(maxSpans) || maxSpans < 0) {
      throw new RangeError('maxSpans must be a nonnegative integer');
    }
    this.start = clock();
  }
  private elapsed(): number { return this.clock() - this.start; }
  setupComplete(): void {
    if (this.setupEnd === null) this.setupEnd = this.elapsed();
  }
  classified<T extends string>(verdict: T): T {
    if ((verdict === 'verified' || verdict === 'refuted') && this.terminal === null) {
      this.terminal = { verdict: verdict === 'verified' ? 'verified' : 'refuted', at_ms: this.elapsed(),
        source: 'existing_fixture_state_classification' };
    }
    return verdict;
  }
  private record(phase: string, start: number, errorType: string | null): void {
    const event = { event: 'span', task_id: this.taskId, clock_id: this.clockId,
      phase, start_ms: start, end_ms: this.elapsed(), error_type: errorType };
    if (this.spans.length < this.maxSpans) this.spans.push(event);
    else this.dropped += 1;
  }
  measureSync<T>(phase: string, operation: () => T): T {
    const start = this.elapsed();
    let errorType: string | null = null;
    try { return operation(); }
    catch (error: unknown) {
      errorType = error instanceof Error ? error.name : 'UnknownError';
      throw error;
    } finally { this.record(phase, start, errorType); }
  }
  async measureAsync<T>(phase: string, operation: () => Promise<T>): Promise<T> {
    const start = this.elapsed();
    let errorType: string | null = null;
    try { return await operation(); }
    catch (error: unknown) {
      errorType = error instanceof Error ? error.name : 'UnknownError';
      throw error;
    } finally { this.record(phase, start, errorType); }
  }
  finish(outcome: string, dryRun = false, errorType: string | null = null): Event[] {
    if (this.closed) throw new Error('task timing already closed');
    this.closed = true;
    const end = this.elapsed();
    return [...this.spans, {
      event: 'task_root', schema_version: 2, task_id: this.taskId, clock_id: this.clockId,
      start_ms: 0, end_ms: end, setup_ms: this.setupEnd, outcome,
      outcome_source: 'recipe_return_not_independent_oracle', error_type: errorType,
      dry_run: dryRun, dropped_spans: this.dropped,
      scope: 'run_entry_through_context_cleanup_before_timing_flush',
      terminal_observation: this.terminal,
      post_terminal_observation_ms: this.terminal === null ? null : end - this.terminal.at_ms,
    }];
  }
}

File: libs/cua-driver/examples/jev-use/experiments/task-timing/task_timing.py
"""Recipe-local timing; never an independent success oracle or Driver contract."""
from __future__ import annotations

from collections.abc import Callable, Iterator
from contextlib import contextmanager
import time
import uuid
from typing import Any


class TaskTiming:
    def __init__(self, *, clock: Callable[[], int] = time.perf_counter_ns,
                 max_spans: int = 512) -> None:
        if max_spans < 0:
            raise ValueError("max_spans must be nonnegative")
        self.task_id = uuid.uuid4().hex
        self.clock_id = f"perf_counter_ns:{self.task_id}"
        self._clock = clock
        self._start = clock()
        self._spans: list[dict[str, Any]] = []
        self._max_spans = max_spans
        self._dropped = 0
        self._setup_end: float | None = None
        self._closed = False

    def _elapsed(self) -> float:
        return (self._clock() - self._start) / 1_000_000

    def setup_complete(self) -> None:
        if self._setup_end is None:
            self._setup_end = self._elapsed()

    @contextmanager
    def span(self, phase: str) -> Iterator[None]:
        start = self._elapsed()
        error_type = None
        try:
            yield
        except BaseException as error:
            error_type = type(error).__name__
            raise
        finally:
            event = {
                "event": "span", "task_id": self.task_id, "clock_id": self.clock_id,
                "phase": phase, "start_ms": start, "end_ms": self._elapsed(),
                "error_type": error_type,
            }
            if len(self._spans) < self._max_spans:
                self._spans.append(event)
            else:
                self._dropped += 1

    def finish(self, outcome: str, *, dry_run: bool = False,
               error_type: str | None = None) -> list[dict[str, Any]]:
        if self._closed:
            raise RuntimeError("task timing already closed")
        self._closed = True
        end = self._elapsed()
        return [*self._spans, {
            "event": "task_root", "task_id": self.task_id, "clock_id": self.clock_id,
            "start_ms": 0.0, "end_ms": end, "setup_ms": self._setup_end,
            "outcome": outcome, "outcome_source": "recipe_return_not_independent_oracle",
            "error_type": error_type, "dry_run": dry_run,
            "dropped_spans": self._dropped,
            "scope": "run_entry_through_context_cleanup_before_timing_flush",
        }]
